iree-prof-tools/graph-util.cc

```diff
@@ -0,0 +1,428 @@
+  return input;
+}
+
+// Adds an output with some default metadata into |node|.
+Metadata& AddOutputMetadata(GraphNode& node) {
+  Metadata& output = node.outputs_metadata.emplace_back();
+  int output_index = node.outputs_metadata.size() - 1;
+  output.id = absl::StrCat("output-", output_index);
+  output.attrs.emplace_back("index", absl::StrCat(output_index));
+  return output;
+}
+
+// Adds a node for |op| into |graph|.
+GraphNode& AddNodeForOperation(
+    mlir::Operation* op,
+    absl::string_view label,
+    const absl::flat_hash_map<mlir::Operation*, std::string>& op_namespaces,
+    Graph& graph) {
+  CHECK(op_namespaces.contains(op->getParentOp()));
+  return AddNode(label, op_namespaces.at(op->getParentOp()), graph);
+}
+
+// Adds a node for |op| with a default label, i.e. op's name into |graph|.
+GraphNode& AddNodeForOperation(
+    mlir::Operation* op,
+    const absl::flat_hash_map<mlir::Operation*, std::string>& op_namespaces,
+    Graph& graph) {
+  return AddNodeForOperation(op, ToStr(op->getName()), op_namespaces, graph);
+}
+
+// Finds a node whose label is matched with |label_to_match|.
+GraphNode* FindNodeForLabel(const Graph& graph,
+                            absl::string_view label_to_match) {
+  for (const auto& n : graph.nodes) {
+    if (label_to_match == n->label) {
+      return n.get();
+    }
+  }
+  return nullptr;
+}
+
+// Records op name as a namespace for ops which are just containers from graph's
+// point of view.
+void AddNamespace(
+    mlir::Operation* op,
+    absl::flat_hash_map<mlir::Operation*, std::string>& op_namespaces) {
+  CHECK(op_namespaces.contains(op->getParentOp()));
+  int index = op_namespaces.size();
+  op_namespaces[op] = absl::StrCat(op_namespaces[op->getParentOp()], "/",
+                                   ToStr(op->getName()), "-", index);
+}
+
+// Adds a |resource| of stream.cmd.dispatch |op| as an output or an incoming
+// edge of |node|.
+void AddResourceAsOutputOrIncomingEdge(IREE::Stream::CmdDispatchOp op,
+                                       int resource_index,
+                                       const Graph& graph,
+                                       GraphNode& node) {
+  auto resource =
+      llvm::cast<mlir::BlockArgument>(op.getResources()[resource_index]);
+  int index = resource.getArgNumber();
+  int64_t offset = ToInt(op.getResourceOffsets()[resource_index]);
+  int64_t size = ToInt(op.getResourceLengths()[resource_index]);
+  auto resource_id = GetResourceId(index, offset, size);
+
+  auto access = ToAccess(op.getResourceAccesses()[resource_index]);
+  if (access & ACCESS_WRITE) {
+    auto& output = AddOutputMetadata(node);
+    output.id = resource_id;
+    output.attrs.emplace_back("arg_index", absl::StrCat(index));
+    output.attrs.emplace_back("offset", absl::StrCat(offset));
+    output.attrs.emplace_back("size", absl::StrCat(size));
+    output.attrs.emplace_back("type", ToStr(resource.getType()));
+    // Treat ACCESS_READ_WRITE as ACCESS_WRITE.
```

Comment:
I think I understand the situation and it is not great since there is a case that a resource is used for an input and output.

Is there a restriction when using a node as input/output? It will create a cycle though.